### PR TITLE
doc/mgr: Enhance placement specs 

### PR DESCRIPTION
**Pull request by sebastian-philipp**
relates to #33924

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/mgr/orchestrator.rst [hunk 1]
```diff
@@ -273,11 +273,72 @@ Placement Specification
...
+  
+    service_type: prometheus
+    placement:
+      hosts: "host1 host2 host3"
+     
+MONs and other services may require some enhanced network specifications::
+
+  orch daemon add mon myhost:[v2:1.2.3.4:3000,v1:1.2.3.4:6789]=name
+  
+Where ``[v2:1.2.3.4:3000,v1:1.2.3.4:6789]`` is the netowrk address of the monitor
```
Comment (liewegas) [inline doc/mgr/orchestrator.rst]: network

File: doc/mgr/orchestrator.rst [hunk 2]
```diff
@@ -273,11 +273,72 @@ Placement Specification
...
+and ``=name`` will specify the name of the new monitor.
+
+Placement by Labels
+^^^^^^^^^^^^^^^^^^^
+
+Daemons can be explictly placed on hosts that match a specifc label::
+
+    orch apply prometheus "label:mylabel"
```
Comment (liewegas) [inline doc/mgr/orchestrator.rst]: quotes aren't needed here?

File: doc/mgr/orchestrator.rst [hunk 3]
```diff
@@ -273,11 +273,72 @@ Placement Specification
...
+    placement:
+      label: "mylabel"
+
+
+Placement on all hosts
+^^^^^^^^^^^^^^^^^^^^^^
+
+Daemons can be explictly placed on all hosts as well::
+
+    orch apply prometheus "all:true"
```
Comment (liewegas) [inline doc/mgr/orchestrator.rst]: or here

File: doc/mgr/orchestrator.rst [hunk 4]
```diff
@@ -273,11 +273,72 @@ Placement Specification
...
+Setting a limit
+^^^^^^^^^^^^^^^
+
+By specifying ``count``, only that number of daemons will be created::
+
+    orch apply prometheus 3
+    
+To deploy *daemons* on a subset of hosts, also specify the count::
+
+    orch apply prometheus "2 host1 host2 host3"
```
Comment (liewegas) [inline doc/mgr/orchestrator.rst]: could also add the converse, e.g., "3 host1 host2"

File: doc/mgr/orchestrator.rst [hunk 5]
```diff
@@ -273,11 +273,72 @@ Placement Specification
...
+
+or in yaml::
+
+    service_type: prometheus
+    placement:
+      count: 3
```
Comment (liewegas) [inline doc/mgr/orchestrator.rst]: show a yaml example with hosts too?

File: doc/mgr/orchestrator.rst (hunk 1 above)
Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: fixed.

File: doc/mgr/orchestrator.rst (hunk 2 above)
Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: fixed.

File: doc/mgr/orchestrator.rst (hunk 3 above)
Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: fixed.

File: doc/mgr/orchestrator.rst (hunk 4 above)
Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: fixed.

File: doc/mgr/orchestrator.rst (hunk 5 above)
Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: fixed.

**Comment by liewegas** [comment]
```
/home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/lib/python3.6/site-packages/sphinx/util/docutils.py:287: RemovedInSphinx30Warning: function based directive support is now deprecated. Use class based directive instead.
  RemovedInSphinx30Warning)
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/mgr/orchestrator.rst:278: WARNING: Title level inconsistent:

Explicit placements
^^^^^^^^^^^^^^^^^^^
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/mgr/orchestrator.rst:298: WARNING: Title level inconsistent:

Placement by Labels
^^^^^^^^^^^^^^^^^^^
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/mgr/orchestrator.rst:312: WARNING: Title level inconsistent:

Placement on all hosts
^^^^^^^^^^^^^^^^^^^^^^
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/mgr/orchestrator.rst:326: WARNING: Title level inconsistent:

Setting a limit
^^^^^^^^^^^^^^^
```

File: doc/mgr/orchestrator.rst [hunk 6]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
-* To deploy *daemons* on specific hosts, specify the host names: ``host1 host2 host3``
-* To deploy *daemons* on a subset of hosts, also specify the count: ``2 host1 host2 host3``
+Explicit placements
+^^^^^^^^^^^^^^^^^^^
 
+Daemons can be explictly placed on hosts by simply specifying them::
+
+    orch apply prometheus "host1 host2 host3"
+    
+or in yaml::
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
Or in yaml::

File: doc/mgr/orchestrator.rst [hunk 7]
```diff
@@ -273,13 +273,85 @@ Placement Specification
 
 In order to allow the orchestrator to deploy a *service*, it needs to
 know how many and where it should deploy *daemons*. The orchestrator 
-defines a placement specification:
+defines a placement specification that can either be passed as command line arguments.
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]: Is this meant to be "can either be passed as .... or...." ? are you missing a second half here? If not, see my suggestion below

Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
defines a placement specification that can be passed as a command line argument.

File: doc/mgr/orchestrator.rst [hunk 8]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+    service_type: prometheus
+    placement:
+      hosts: "host1 host2 host3"
+     
+MONs and other services may require some enhanced network specifications::
+
+  orch daemon add mon myhost:[v2:1.2.3.4:3000,v1:1.2.3.4:6789]=name
+  
+Where ``[v2:1.2.3.4:3000,v1:1.2.3.4:6789]`` is the network address of the monitor
+and ``=name`` will specify the name of the new monitor.
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
and ``=name`` specifies the name of the new monitor.

File: doc/mgr/orchestrator.rst [hunk 9]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+
+Placement by Labels
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
Placement by labels

File: doc/mgr/orchestrator.rst [hunk 10]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+^^^^^^^^^^^^^^^^^^^
+
+Daemons can be explictly placed on hosts that match a specifc label::
+
+    orch apply prometheus label:mylabel
+
+or in yaml::
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
``` suggestion
Or in yaml::

File: doc/mgr/orchestrator.rst [hunk 11]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+
+
+Placement on all hosts
+^^^^^^^^^^^^^^^^^^^^^^
+
+Daemons can be explictly placed on all hosts as well::
+
+    orch apply prometheus all:true
+
+or in yaml::
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
``` suggestion
Or in yaml::

File: doc/mgr/orchestrator.rst [hunk 12]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+
+By specifying ``count``, only that number of daemons will be created::
+
+    orch apply prometheus 3
+    
+To deploy *daemons* on a subset of hosts, also specify the count::
+
+    orch apply prometheus "2 host1 host2 host3"
     
+If the count is bigger than the amount of hosts, cephadm will still deploy two daemons::
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
If the count is bigger than the amount of hosts, cephadm still deploys two daemons::

File: doc/mgr/orchestrator.rst [hunk 13]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+
+    orch apply prometheus "3 host1 host2"
+
+or in yaml::
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
Or in yaml::

File: doc/mgr/orchestrator.rst [hunk 14]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
+
+    service_type: prometheus
+    placement:
+      count: 3
+      
+or with hosts::
```
Comment (asettle) [inline doc/mgr/orchestrator.rst]:
```suggestion
Or with hosts::

File: doc/mgr/orchestrator.rst [hunk 15]
```diff
@@ -273,13 +273,85 @@ Placement Specification
...
 
+Daemons can be explictly placed on hosts by simply specifying them::
+
+    orch apply prometheus "host1 host2 host3"
+    
+or in yaml::
+  
+    service_type: prometheus
+    placement:
+      hosts: "host1 host2 host3"
```
Comment (jschmid1) [inline doc/mgr/orchestrator.rst]: I wonder if we want to support both syntactical versions of `hosts` definitions in the future.

```yaml
    service_type: prometheus
    placement:
      hosts: "host1 host2 host3"
```

vs

```yaml
    service_type: prometheus
    placement:
      hosts: 
        - host1
        - host2
        - host3
```

Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: > I wonder if we want to support both syntactical versions of `hosts` definitions in the future.
> 
> ```yaml
>     service_type: prometheus
>     placement:
>       hosts: "host1 host2 host3"
> ```

Indeed this doesn't really make sense. Either we parse the whole PlacemetSpec or we don't. 

```yaml
service_type: prometheus
placement: "3 label:monitoring"
```

or 

```yaml
service_type: prometheus
placement: 
  count: 3
  label: "monitoring"
  hosts:
    - host1
    - "host2:2001:0db8:85a3:0000:0000:8a2e:0370:733=yougettheidea"
```

I'm ok with both!

Comment (jschmid1) [reply doc/mgr/orchestrator.rst]: or we recommend a very explicit form in the first place:

this is how we internally save the placement struct if a `hosts` struct was given:

``` yaml
placement:
  hosts:
  - hostname: cephadm-dev
    name: ''
    network: ''
```

A bit more verbose, but it's very descriptive. I can imagine people struggling with the syntax we used on the commandline (host:network=optional_name).

Comment (sebastian-philipp) [reply doc/mgr/orchestrator.rst]: imo too verbose :smile:

**Pull request merged**